assets/predictive-search.js

```diff
@@ -0,0 +1,201 @@
+class PredictiveSearch extends HTMLElement {
+  constructor() {
+    super();
+    this.cachedResults = {};
+    this.input = this.querySelector('input[type="search"]');
+    this.predictiveSearchResults = this.querySelector('[data-predictive-search]');
+
+    this.setupEventListeners();
+  }
+
+  setupEventListeners() {
+    const form = this.querySelector('form.search');
+    form.addEventListener('submit', this.onFormSubmit.bind(this));
+
+    this.input.addEventListener('input', debounce((event) => {
+      this.onChange(event);
+    }, 300).bind(this));
+    this.input.addEventListener('focus', this.onFocus.bind(this));
+
+    this.addEventListener('focusout', this.onFocusOut.bind(this));
+    this.addEventListener('keyup', this.onKeyup.bind(this));
+    this.addEventListener('keydown', this.onKeydown.bind(this));
+  }
+
+  getQuery() {
+    return this.input.value.trim();
+  }
+
+  onChange() {
+    const searchTerm = this.getQuery();
+
+    if (!searchTerm.length) {
+      this.close(true);
+      return;
+    }
+
+    this.getSearchResults(searchTerm);
+  }
+
+  onFormSubmit(event) {
+    if (!this.getQuery().length || this.querySelector('[aria-selected="true"] a')) event.preventDefault();
+  }
+
+  onFocus() {
+    const searchTerm = this.getQuery();
+
+    if (!searchTerm.length) return;
+
+    if (this.getAttribute('results') === 'true') {
+      this.open();
+    } else {
+      this.getSearchResults(searchTerm);
+    }
+  } 
+
+  onFocusOut() {
+    setTimeout(() => {
+      if (!this.contains(document.activeElement)) this.close();
```

Comment:
Does it make sense on mobile (tested on iOS) to have the search results close when the on-screen keyboard is dismissed?
![Kapture 2021-08-30 at 13 56 08](https://user-images.githubusercontent.com/3710548/131384595-553ee176-9b8a-486c-8dcd-5a68d4d9d3e1.gif)